### DISCUSSION TITLE
commitlint, husky & commitizen

### DISCUSSION THREAD

**Discussion by RicardoJBarrios** [jscutlery/semver#197] [Ideas]
Hello.

I am researching to add these dependencies, `commitlint`, `husky` and `commitizen`, optionally when installing the package, and I have thought about adding the following parameters to `ngAdd`:

| Flag | Description|
| --- | --- |
| `--skipInstall` | Skips installing new `node_modules` after modifying `package.json` |
| `--packages="commitlint, husky, commitizen"` | Define the packages to add. |
| `--overrideConfigurations` | Do override existing configuration parameters if necesary |
| `--force` | Ignore errors and override already existing files |

If `--packages` is not specified, none of the dependencies or configurations will be installed.

What do you think about this options? Should the library install them by default?

**Comment by edbzn** [comment]
Hello @RicardoJBarrios, 

Agree for adding a `--skip-install` option. :+1: 

I'm wondering how `--override configurations` works because there are multiple configurations, why not just providing default configuration for these packages? In this case, I'm not sure about the `--force` option. What do you think?

I would suggest using a prompt to ask which packages the user wants to install, by default all dependencies are checked, and also provide an option `--packages` as you suggested to skip the prompt.

**Reply by RicardoJBarrios** [reply]
The idea of ​​the override setting is to avoid conflicts with previous configs. If the user installs this library and has a previous configuration, she can choose to leave the old configuration or restart it with the defaults established by this library.

If we do not implement `override`, the behavior should be to do not override each library config if exists and setup the defaults if does not exist. Maybe it's a best option.

I agree with the prompt, but maybe add a `"none"` option to `--packages` to avoid the install without ask.

**Reply by edbzn** [reply]
> If we do not implement override, the behavior should be to do not override each library config if exists and setup the defaults if does not exist. Maybe it's a best option.

I got it now, I'm in favor of not overriding configs in case they are already defined. I think the best would be a prompt in case a conflict was detected to ask if the user wants to overide, but it's certainly not essential to start, let's go for your suggestion for now.

> I agree with the prompt, but maybe add a "none" option to --packages to avoid the install without ask.

Agree about this.

**Reply by RicardoJBarrios** [reply]
Perfect. I'll work with this options.

**Comment by yjaaidi** [comment]
Hey @RicardoJBarrios , thanks for the suggestion and for your time! 

I'm generally in favor of reducing the surface and minimizing options and prompts.

I'd say that commitlint, husky and commitizen are implementation details.

What about an opt-in like `--enforce-conventional-commits`  that adds the packages + the configuration if not already there.
As mentioned by @edbzn , if the configuration is already present, then ppl know what they are doing, so let them do so 😊

For the minority of ppl who would want just commitizen or commitlint etc... then just have to edit what was generated and removed whatever they want to remove.

We end up with just `--skip-install` and `--enforce-conventional-commits`.

What do you think?

**Reply by edbzn** [reply]
> What about an opt-in like --enforce-conventional-commits that adds the packages + the configuration if not already there.

I like this suggestion!

**Reply by RicardoJBarrios** [reply]
I also like the idea. I will do it that way.

**Reply by yjaaidi** [reply]
Thank you so much Ricardo 🙏

You can't imagine enough how it feels to get feedback and help like you do.

Let us know if you need any help.
Don't hesitate to share your dev branch link with us so we can discuss it or even contribute 😊.

**Reply by RicardoJBarrios** [reply]
You're wellcome.

I usually create a lot of projects in my work, and I always set something like that. I was also looking to learn how to create these types of plugins so glad I could help.

I'll do it in https://github.com/RicardoJBarrios/semver/tree/commitlint-husky-commitizen.

**Comment by RicardoJBarrios** [comment]
Now I have something working in https://github.com/RicardoJBarrios/semver/tree/commitlint-husky-commitizen, but I don't know if there is a better or standard way to install this deps and set the configs. If you think it's ok I'll do the tests and I'll create a PR.

**Reply by edbzn** [reply]
Looks good to me, well done! You can open a PR we will make feedback on it.
